src/cpu/aarch64/acl_winograd_convolution.hpp

```diff
@@ -59,7 +59,7 @@ struct acl_wino_convolution_fwd_t : public primitive_t {
 private:
     status_t execute_forward(const exec_ctx_t &ctx) const;
     const pd_t *pd() const { return (const pd_t *)primitive_t::pd().get(); }
-    std::unique_ptr<acl_obj_t<Op>> acl_obj_;
+    mutable std::unique_ptr<acl_obj_t<Op>> acl_obj_;
```

Comment:
why not declare acl_obj_ directly in execute forward?
IIUC, this patch would still break in cache of concurrent execution.

Reply:
Thanks. I am still confused by the use of the mutex. Why not make the object local to the execute function call? Each execute call would have his own object and this will remove any need for synchronization.

Furthermore, as currently implemented, I believe you might still have an issue in case of concurrent execution as only the initialization is mutex protected. For example if you have
```
t0: lock -> acl_obj init *, dtor previous -> unlock ->  acl_obj.get() ->            ...             -> using handle
t1:                                                     lock          -> acl_obj init, dtor previous *-> unlock -> acl_obj.get() -> ... 
```

In the above example, t0 will get some pointer to object upon acl_obj.get() call, but the underlying object will be destroyed by t1 acl_obj init.

Reply:
hi, @mgouicem, thanks for your help, i've changed the acl_obj to a local variable. 

FYI, I'm kind of confused about what kind of synchronisation should be posed on oneDNN primitives and how the primitives cache works? could you please share some docs/info?

Reply:
Regarding synchronization, primitive.execute function should be const qualified. So because it is not supposed to change any state of the primitive object, there should be no sync necessary in execution.

We do have synchronizations happening upon primitive cache accesses (single writer, multiple readers). This should be transparent from primitive implementation.

Regarding scratchpad, when user managed, scratchpad is just another output from a primitive perspective (so user responsibility to sync when needed). When library managed, it is trickier as it is not thread-safe by default. You can find more info in the scratchpad devguide [page](https://oneapi-src.github.io/oneDNN/dev_guide_attributes_scratchpad.html).

Reply:
thanks mgouicem, that's really helpful!